### PR TITLE
Orbit fixes for MC

### PR DESCRIPTION
**Pull request by bkueng**
This fixes a few problems for orbit:
remove command ack for VEHICLE_CMD_DO_ORBIT. Already acked in commander

Fixes a regression from https://github.com/PX4/PX4-Autopilot/commit/8bae4e5c0e29f70d5d1d2427ffdef97092be939c, where
the orbit flight task wasn't an extra task (flight_tasks_to_add) anymore
and therefore the command handling wasn't generated.

There was a race condition that could cause several outcomes. The most severe
was that flight_mode_manager gets the command, switches to orbit and then
in the next iteration switches back because commander did not change
nav_state yet. When commander then switches, flight_mode_manager would still
be in the old mode.
This is prevented by storing the command (allowing it to arrive before or
after mode switch), and then apply it after the switch happens.

### CODE REVIEW

**Comment by dagar** [comment]
We've had similar issues in other areas as well (eg CMD_DO_REPOSITION, CMD_NAV_TAKEOFF, etc), maybe we could do something like split the commands internally and stop tolerating these vehicle_command races in the first place?

Maybe something to tackle on top of services later.

**Comment by bkueng** [comment]
Yeah I was also thinking about that. But even if you guarantee publication ordering there's no guarantee the receiver sees it in that order as well, unless the logic on the receiver side is specifically arranged to handle that. Both ways are possible though.

**Comment by dagar** [comment]
> Yeah I was also thinking about that. But even if you guarantee publication ordering there's no guarantee the receiver sees it in that order as well, unless the logic on the receiver side is specifically arranged to handle that. Both ways are possible though.

I meant doing it more explicitly with either mavlink or commander owning the external command `MAV_CMD_DO_ORBIT` and handling the mode change and then sending (/forwarding) a new dedicated internal orbit command/service that only the flight task respects.

Yet another idea would be factoring in FlightTask/FlightMode specific minimal command/service handling when the mode is dormant, but I'm not sure if this is a sufficiently common use case we need to worry about.

Either way probably not worth getting into right now because once services are in place I think we'll want to prevent MAV_CMDs from leaking into the wider system and instead bridge all of this stuff between MAV_CMD <=> PX4 srv.

**Review by dagar — APPROVED** [review]

**Comment by bkueng** [comment]
> I meant doing it more explicitly with either mavlink or commander owning the external command MAV_CMD_DO_ORBIT and handling the mode change and then sending (/forwarding) a new dedicated internal orbit command/service that only the flight task respects.

Yes we're talking about the same thing. The race condition happens if the flight mode manager is right before reading the command, commander publishes both topics, then the manager will see the command before the mode change. One way to avoid it is to let the receiver check for commands, then check & handle the mode change, then handle the command.

> Either way probably not worth getting into right now because once services are in place I think we'll want to prevent MAV_CMDs from leaking into the wider system and instead bridge all of this stuff between MAV_CMD <=> PX4 srv.

I think so too

File: src/modules/flight_mode_manager/FlightModeManager.cpp [hunk 1]
```diff
@@ -265,47 +280,39 @@ void FlightModeManager::start_flight_task()
...
+void FlightModeManager::tryApplyCommandIfAny()
+{
+	if (isAnyTaskActive() && _current_command.command != 0 && hrt_absolute_time() < _current_command.timestamp + 200_ms) {
+		bool success = false;
+
+		if (_current_task.task->applyCommandParameters(_current_command, success)) {
+			_current_command.command = 0;
+
+			if (!success) {
+				switchTask(FlightTaskIndex::Failsafe);
```
Comment (MaEtUgR) [inline src/modules/flight_mode_manager/FlightModeManager.cpp:292]: This is the tricky part where we need feedback to the state machine.

**Review by MaEtUgR — APPROVED** [review]
Thanks a lot for looking into this and coming up with an improvement over my reliance on timing.

**Pull request merged**